File: b2b-marketplace-complete/scripts/test_api.py
#!/usr/bin/env python3
"""
B2B Marketplace API Test Script
================================
Script Python để test tất cả API endpoints

Cài đặt:
    pip install requests rich websocket-client

Sử dụng:
    python test_api.py              # Chạy tất cả tests
    python test_api.py --auth       # Test authentication only
    python test_api.py --products   # Test products only
    python test_api.py --interactive # Chế độ interactive
"""

import requests
import json
import argparse
import sys
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from datetime import datetime

# Try to import rich for better output
try:
    from rich.console import Console
    from rich.table import Table
    from rich.panel import Panel
    from rich import print as rprint
    RICH_AVAILABLE = True
    console = Console()
except ImportError:
    RICH_AVAILABLE = False
    console = None


# ==================== CONFIGURATION ====================

@dataclass
class Config:
    base_url: str = "http://localhost:3001/api"
    timeout: int = 30
    
    # Test accounts
    admin_email: str = "admin@b2bmarket.com"
    admin_password: str = "Admin123!@#"
    
    supplier_email: str = "supplier@techcorp.com"
    supplier_password: str = "Supplier123!@#"
    
    shop_email: str = "buyer@retailplus.com"
    shop_password: str = "Shop123!@#"


config = Config()


# ==================== API CLIENT ====================

class B2BApiClient:
    """Client để tương tác với B2B Marketplace API"""
    
    def __init__(self, base_url: str = None):
        self.base_url = base_url or config.base_url
        self.access_token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        self.user: Optional[Dict] = None
        self.session = requests.Session()
    
    def _headers(self, auth: bool = True) -> Dict[str, str]:
        """Tạo headers cho request"""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if auth and self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        return headers
    
    def _request(
        self, 
        method: str, 
        endpoint: str, 
        data: Dict = None, 
        params: Dict = None,
        auth: bool = True
    ) -> Dict[str, Any]:
        """Thực hiện HTTP request"""
        url = f"{self.base_url}{endpoint}"
        headers = self._headers(auth)
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                headers=headers,
                json=data,
                params=params,
                timeout=config.timeout
            )
            
            result = response.json()
            result['_status_code'] = response.status_code
            return result
            
        except requests.exceptions.ConnectionError:
            return {
                "success": False,
                "error": {"message": "Connection failed. Is the server running?"},
                "_status_code": 0
            }
        except requests.exceptions.Timeout:
            return {
                "success": False,
                "error": {"message": "Request timed out"},
                "_status_code": 0
            }
        except Exception as e:
            return {
                "success": False,
                "error": {"message": str(e)},
                "_status_code": 0
            }
    
    # ==================== AUTH ====================
    
    def login(self, email: str, password: str) -> Dict:
        """Đăng nhập và lưu tokens"""
        result = self._request(
            "POST", "/auth/login",
            {"email": email, "password": password},
            auth=False
        )
        
        if result.get("success"):
            self.access_token = result["data"]["accessToken"]
            self.refresh_token = result["data"]["refreshToken"]
            self.user = result["data"]["user"]
        
        return result
    
    def register(self, email: str, password: str, role: str, **kwargs) -> Dict:
        """Đăng ký tài khoản mới"""
        data = {
            "email": email,
            "password": password,
            "role": role,
            **kwargs
        }
        return self._request("POST", "/auth/register", data, auth=False)
    
    def get_me(self) -> Dict:
        """Lấy thông tin user hiện tại"""
        return self._request("GET", "/auth/me")
    
    def refresh_tokens(self) -> Dict:
        """Refresh access token"""
        result = self._request(
            "POST", "/auth/refresh",
            {"refreshToken": self.refresh_token},
            auth=False
        )
        
        if result.get("success"):
            self.access_token = result["data"]["accessToken"]
            self.refresh_token = result["data"]["refreshToken"]
        
        return result
    
    def logout(self) -> Dict:
        """Đăng xuất"""
        result = self._request(
            "POST", "/auth/logout",
            {"refreshToken": self.refresh_token}
        )
        if result.get("success"):
            self.access_token = None
            self.refresh_token = None
            self.user = None
        return result
    
    # ==================== PRODUCTS ====================
    
    def list_products(
        self, 
        page: int = 1, 
        limit: int = 20,
        search: str = None,
        category_id: str = None,
        min_price: float = None,
        max_price: float = None
    ) -> Dict:
        """Lấy danh sách sản phẩm"""
        params = {"page": page, "limit": limit}
        if search:
            params["search"] = search
        if category_id:
            params["categoryId"] = category_id
        if min_price:
            params["minPrice"] = min_price
        if max_price:
            params["maxPrice"] = max_price
            
        return self._request("GET", "/products", params=params, auth=False)
    
    def get_product(self, product_id: str) -> Dict:
        """Lấy chi tiết sản phẩm"""
        return self._request("GET", f"/products/{product_id}", auth=False)
    
    def create_product(self, data: Dict) -> Dict:
        """Tạo sản phẩm mới (Supplier only)"""
        return self._request("POST", "/products", data)
    
    def update_product(self, product_id: str, data: Dict) -> Dict:
        """Cập nhật sản phẩm"""
        return self._request("PUT", f"/products/{product_id}", data)
    
    def delete_product(self, product_id: str) -> Dict:
        """Xóa sản phẩm"""
        return self._request("DELETE", f"/products/{product_id}")
    
    def get_featured_products(self) -> Dict:
        """Lấy sản phẩm nổi bật"""
        return self._request("GET", "/products/featured", auth=False)
    
    # ==================== SUPPLIERS ====================
    
    def list_suppliers(self, page: int = 1, limit: int = 20) -> Dict:
        """Lấy danh sách nhà cung cấp"""
        return self._request(
            "GET", "/suppliers", 
            params={"page": page, "limit": limit}, 
            auth=False
        )
    
    def get_supplier(self, supplier_id: str) -> Dict:
        """Lấy chi tiết nhà cung cấp"""
        return self._request("GET", f"/suppliers/{supplier_id}", auth=False)
    
    def get_my_supplier_profile(self) -> Dict:
        """Lấy profile supplier của mình"""
        return self._request("GET", "/suppliers/me/profile")
    
    def get_supplier_stats(self, supplier_id: str) -> Dict:
        """Lấy thống kê của supplier"""
        return self._request("GET", f"/suppliers/{supplier_id}/stats", auth=False)
    
    # ==================== CATEGORIES ====================
    
    def get_categories(self) -> Dict:
        """Lấy cây danh mục"""
        return self._request("GET", "/categories", auth=False)
    
    def get_category_by_slug(self, slug: str) -> Dict:
        """Lấy danh mục theo slug"""
        return self._request("GET", f"/categories/slug/{slug}", auth=False)
    
    # ==================== NEGOTIATIONS ====================
    
    def list_negotiations(self, page: int = 1, limit: int = 20) -> Dict:
        """Lấy danh sách đàm phán"""
        return self._request("GET", "/negotiations", params={"page": page, "limit": limit})
    
    def get_negotiation(self, negotiation_id: str) -> Dict:
        """Lấy chi tiết đàm phán"""
        return self._request("GET", f"/negotiations/{negotiation_id}")
    
    def create_negotiation(
        self, 
        product_id: str, 
        supplier_id: str, 
        message: str
    ) -> Dict:
        """Tạo phiên đàm phán mới (Shop only)"""
        return self._request("POST", "/negotiations", {
            "productId": product_id,
            "supplierId": supplier_id,
            "initialMessage": message
        })
    
    def send_message(
        self, 
        negotiation_id: str, 
        content: str, 
        msg_type: str = "TEXT"
    ) -> Dict:
        """Gửi tin nhắn trong phiên đàm phán"""
        return self._request("POST", f"/negotiations/{negotiation_id}/messages", {
            "type": msg_type,
            "content": content
        })
    
    def get_messages(self, negotiation_id: str, page: int = 1) -> Dict:
        """Lấy tin nhắn của phiên đàm phán"""
        return self._request(
            "GET", 
            f"/negotiations/{negotiation_id}/messages",
            params={"page": page}
        )
    
    # ==================== PURCHASE INTENTS ====================
    
    def list_purchase_intents(self, page: int = 1, limit: int = 20) -> Dict:
        """Lấy danh sách purchase intents"""
        return self._request(
            "GET", "/purchase-intents", 
            params={"page": page, "limit": limit}
        )
    
    def create_purchase_intent(self, data: Dict) -> Dict:
        """Tạo purchase intent mới"""
        return self._request("POST", "/purchase-intents", data)
    
    # ==================== HEALTH ====================
    
    def health_check(self) -> Dict:
        """Kiểm tra health của server"""
        try:
            response = self.session.get(
                f"{self.base_url.replace('/api', '')}/health",
                timeout=5
            )
            return {"success": True, "status": response.text}
        except:
            return {"success": False, "status": "Server not reachable"}
    
    def health_ready(self) -> Dict:
        """Kiểm tra DB và Redis"""
        try:
            response = self.session.get(
                f"{self.base_url.replace('/api', '')}/health/ready",
                timeout=5
            )
            return response.json()
        except Exception as e:
            return {"success": False, "error": str(e)}


# ==================== TEST RUNNER ====================

class TestRunner:
    """Runner để chạy các test cases"""
    
    def __init__(self):
        self.client = B2BApiClient()
        self.results: List[Dict] = []
        self.passed = 0
        self.failed = 0
    
    def print_header(self, text: str):
        """In header"""
        print("\n" + "=" * 60)
        print(f"  {text}")
        print("=" * 60)
    
    def print_test(self, name: str, success: bool, details: str = ""):
        """In kết quả test"""
        status = "✅ PASS" if success else "❌ FAIL"
        print(f"{status} | {name}")
        if details and not success:
            print(f"       └─ {details}")
        
        self.results.append({
            "name": name,
            "success": success,
            "details": details
        })
        
        if success:
            self.passed += 1
        else:
            self.failed += 1
    
    def run_test(self, name: str, test_func):
        """Chạy một test case"""
        try:
            result = test_func()
            success = result.get("success", False)
            details = "" if success else result.get("error", {}).get("message", "Unknown error")
            self.print_test(name, success, details)
            return result
        except Exception as e:
            self.print_test(name, False, str(e))
            return {"success": False, "error": {"message": str(e)}}
    
    def print_summary(self):
        """In tổng kết"""
        print("\n" + "=" * 60)
        print("  TEST SUMMARY")
        print("=" * 60)
        print(f"  Total:  {self.passed + self.failed}")
        print(f"  Passed: {self.passed} ✅")
        print(f"  Failed: {self.failed} ❌")
        print("=" * 60)
    
    # ==================== TEST SUITES ====================
    
    def test_health(self):
        """Test health endpoints"""
        self.print_header("HEALTH CHECK")
        
        self.run_test("Health Check", lambda: self.client.health_check())
        self.run_test("Health Ready", lambda: self.client.health_ready())
    
    def test_auth(self):
        """Test authentication"""
        self.print_header("AUTHENTICATION")
        
        # Test login với supplier
        result = self.run_test(
            "Login (Supplier)", 
            lambda: self.client.login(config.supplier_email, config.supplier_password)
        )
        
        if result.get("success"):
            # Test get me
            self.run_test("Get Current User", lambda: self.client.get_me())
            
            # Test refresh token
            self.run_test("Refresh Token", lambda: self.client.refresh_tokens())
            
            # Test logout
            self.run_test("Logout", lambda: self.client.logout())
        
        # Test login với shop
        self.run_test(
            "Login (Shop)", 
            lambda: self.client.login(config.shop_email, config.shop_password)
        )
        
        # Logout để clean up
        self.client.logout()
        
        # Test invalid login
        result = self.run_test(
            "Login (Invalid - should fail)", 
            lambda: self.client.login("invalid@email.com", "wrongpassword")
        )
        # This should fail, so we flip the success
        if not result.get("success"):
            self.passed += 1
            self.failed -= 1
    
    def test_products(self):
        """Test products API"""
        self.print_header("PRODUCTS")
        
        # List products (public)
        result = self.run_test("List Products", lambda: self.client.list_products(limit=5))
        
        products = result.get("data", [])
        if products:
            product_id = products[0]["id"]
            
            # Get single product
            self.run_test(
                "Get Product Detail", 
                lambda: self.client.get_product(product_id)
            )
        
        # Test search
        self.run_test(
            "Search Products", 
            lambda: self.client.list_products(search="phone")
        )
        
        # Test featured
        self.run_test("Get Featured Products", lambda: self.client.get_featured_products())
        
        # Test create product (requires auth)
        self.client.login(config.supplier_email, config.supplier_password)
        
        # Get category first
        categories = self.client.get_categories()
        category_id = None
        if categories.get("data"):
            category_id = categories["data"][0]["id"]
        
        if category_id:
            test_product = {
                "name": f"Test Product {datetime.now().timestamp()}",
                "sku": f"TEST-{int(datetime.now().timestamp())}",
                "shortDescription": "Test product created by API test",
                "description": "Full description of test product",
                "basePrice": 99.99,
                "currency": "USD",
                "moq": 10,
                "categoryId": category_id
            }
            
            create_result = self.run_test(
                "Create Product", 
                lambda: self.client.create_product(test_product)
            )
            
            if create_result.get("success"):
                new_product_id = create_result["data"]["id"]
                
                # Update product
                self.run_test(
                    "Update Product",
                    lambda: self.client.update_product(new_product_id, {"basePrice": 149.99})
                )
                
                # Delete product
                self.run_test(
                    "Delete Product",
                    lambda: self.client.delete_product(new_product_id)
                )
        
        self.client.logout()
    
    def test_suppliers(self):
        """Test suppliers API"""
        self.print_header("SUPPLIERS")
        
        # List suppliers (public)
        result = self.run_test("List Suppliers", lambda: self.client.list_suppliers())
        
        suppliers = result.get("data", [])
        if suppliers:
            supplier_id = suppliers[0]["id"]
            
            # Get single supplier
            self.run_test(
                "Get Supplier Detail",
                lambda: self.client.get_supplier(supplier_id)
            )
            
            # Get stats
            self.run_test(
                "Get Supplier Stats",
                lambda: self.client.get_supplier_stats(supplier_id)
            )
        
        # Test authenticated endpoints
        self.client.login(config.supplier_email, config.supplier_password)
        self.run_test("Get My Profile", lambda: self.client.get_my_supplier_profile())
        self.client.logout()
    
    def test_categories(self):
        """Test categories API"""
        self.print_header("CATEGORIES")
        
        # Get category tree
        result = self.run_test("Get Category Tree", lambda: self.client.get_categories())
        
        categories = result.get("data", [])
        if categories:
            slug = categories[0].get("slug")
            if slug:
                self.run_test(
                    "Get Category by Slug",
                    lambda: self.client.get_category_by_slug(slug)
                )
    
    def test_negotiations(self):
        """Test negotiations API"""
        self.print_header("NEGOTIATIONS")
        
        # Login as shop to create negotiation
        self.client.login(config.shop_email, config.shop_password)
        
        # List negotiations
        self.run_test("List Negotiations", lambda: self.client.list_negotiations())
        
        # Get products and suppliers for creating negotiation
        products = self.client.list_products(limit=1)
        suppliers = self.client.list_suppliers()
        
        if products.get("data") and suppliers.get("data"):
            product = products["data"][0]
            supplier = suppliers["data"][0]
            
            # Create negotiation
            create_result = self.run_test(
                "Create Negotiation",
                lambda: self.client.create_negotiation(
                    product["id"],
                    supplier["id"],
                    "Hello, I'm interested in your product. What's the best price for 100 units?"
                )
            )
            
            if create_result.get("success"):
                negotiation_id = create_result["data"]["id"]
                
                # Get negotiation detail
                self.run_test(
                    "Get Negotiation Detail",
                    lambda: self.client.get_negotiation(negotiation_id)
                )
                
                # Get messages
                self.run_test(
                    "Get Messages",
                    lambda: self.client.get_messages(negotiation_id)
                )
                
                # Send message
                self.run_test(
                    "Send Message",
                    lambda: self.client.send_message(
                        negotiation_id,
                        "Looking forward to your response!"
                    )
                )
        
        self.client.logout()
    
    def run_all(self):
        """Chạy tất cả tests"""
        print("\n" + "🚀 " * 20)
        print("  B2B MARKETPLACE API TEST SUITE")
        print("🚀 " * 20)
        print(f"\nBase URL: {self.client.base_url}")
        print(f"Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        
        self.test_health()
        self.test_auth()
        self.test_categories()
        self.test_suppliers()
        self.test_products()
        self.test_negotiations()
        
        self.print_summary()
        
        return self.failed == 0


# ==================== INTERACTIVE MODE ====================

def interactive_mode():
    """Chế độ interactive để test API"""
    client = B2BApiClient()
    
    print("\n" + "=" * 60)
    print("  B2B MARKETPLACE API - INTERACTIVE MODE")
    print("=" * 60)
    print("\nCommands:")
    print("  login <email> <password>  - Login")
    print("  me                        - Get current user")
    print("  products                  - List products")
    print("  suppliers                 - List suppliers")
    print("  categories                - List categories")
    print("  logout                    - Logout")
    print("  exit                      - Exit")
    print()
    
    while True:
        try:
            cmd = input(">>> ").strip().split()
            if not cmd:
                continue
            
            action = cmd[0].lower()
            
            if action == "exit":
                break
            elif action == "login" and len(cmd) >= 3:
                result = client.login(cmd[1], cmd[2])
                print(json.dumps(result, indent=2))
            elif action == "me":
                result = client.get_me()
                print(json.dumps(result, indent=2))
            elif action == "products":
                result = client.list_products(limit=5)
                print(json.dumps(result, indent=2))
            elif action == "suppliers":
                result = client.list_suppliers()
                print(json.dumps(result, indent=2))
            elif action == "categories":
                result = client.get_categories()
                print(json.dumps(result, indent=2))
            elif action == "logout":
                result = client.logout()
                print(json.dumps(result, indent=2))
            else:
                print("Unknown command. Type 'exit' to quit.")
                
        except KeyboardInterrupt:
            print("\nExiting...")
            break
        except Exception as e:
            print(f"Error: {e}")


# ==================== MAIN ====================

def main():
    parser = argparse.ArgumentParser(description="B2B Marketplace API Test Script")
    parser.add_argument("--url", default="http://localhost:3001/api", help="Base URL")
    parser.add_argument("--auth", action="store_true", help="Test auth only")
    parser.add_argument("--products", action="store_true", help="Test products only")
    parser.add_argument("--suppliers", action="store_true", help="Test suppliers only")
    parser.add_argument("--negotiations", action="store_true", help="Test negotiations only")
    parser.add_argument("--interactive", "-i", action="store_true", help="Interactive mode")
    
    args = parser.parse_args()
    
    if args.url:
        config.base_url = args.url
    
    if args.interactive:
        interactive_mode()
        return
    
    runner = TestRunner()
    
    if args.auth:
        runner.test_health()
        runner.test_auth()
    elif args.products:
        runner.test_health()
        runner.test_products()
    elif args.suppliers:
        runner.test_health()
        runner.test_suppliers()
    elif args.negotiations:
        runner.test_health()
        runner.test_negotiations()
    else:
        success = runner.run_all()
        sys.exit(0 if success else 1)
    
    runner.print_summary()


if __name__ == "__main__":
    main()

File: b2b-marketplace-complete/scripts/test_api.js
#!/usr/bin/env node
/**
 * B2B Marketplace API Test Script (Node.js)
 * ==========================================
 * 
 * Cài đặt:
 *   npm install node-fetch@2
 * 
 * Sử dụng:
 *   node test_api.js              # Chạy tất cả tests
 *   node test_api.js --auth       # Test authentication only
 *   node test_api.js --products   # Test products only
 */

const fetch = require('node-fetch') || globalThis.fetch;

// ==================== CONFIG ====================

const CONFIG = {
  baseUrl: process.env.API_URL || 'http://localhost:3001/api',
  timeout: 30000,
  
  // Test accounts
  admin: {
    email: 'admin@b2bmarket.com',
    password: 'Admin123!@#'
  },
  supplier: {
    email: 'supplier@techcorp.com',
    password: 'Supplier123!@#'
  },
  shop: {
    email: 'buyer@retailplus.com',
    password: 'Shop123!@#'
  }
};

// ==================== API CLIENT ====================

class B2BApiClient {
  constructor(baseUrl = CONFIG.baseUrl) {
    this.baseUrl = baseUrl;
    this.accessToken = null;
    this.refreshToken = null;
    this.user = null;
  }

  _headers(auth = true) {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    };
    if (auth && this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }
    return headers;
  }

  async _request(method, endpoint, data = null, params = null, auth = true) {
    let url = `${this.baseUrl}${endpoint}`;
    
    if (params) {
      const searchParams = new URLSearchParams(params);
      url += `?${searchParams.toString()}`;
    }

    const options = {
      method,
      headers: this._headers(auth),
      timeout: CONFIG.timeout
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    try {
      const response = await fetch(url, options);
      const result = await response.json();
      result._statusCode = response.status;
      return result;
    } catch (error) {
      return {
        success: false,
        error: { message: error.message },
        _statusCode: 0
      };
    }
  }

  // ==================== AUTH ====================

  async login(email, password) {
    const result = await this._request('POST', '/auth/login', { email, password }, null, false);
    
    if (result.success) {
      this.accessToken = result.data.accessToken;
      this.refreshToken = result.data.refreshToken;
      this.user = result.data.user;
    }
    
    return result;
  }

  async register(email, password, role, extraData = {}) {
    return this._request('POST', '/auth/register', { email, password, role, ...extraData }, null, false);
  }

  async getMe() {
    return this._request('GET', '/auth/me');
  }

  async refreshTokens() {
    const result = await this._request('POST', '/auth/refresh', { refreshToken: this.refreshToken }, null, false);
    
    if (result.success) {
      this.accessToken = result.data.accessToken;
      this.refreshToken = result.data.refreshToken;
    }
    
    return result;
  }

  async logout() {
    const result = await this._request('POST', '/auth/logout', { refreshToken: this.refreshToken });
    if (result.success) {
      this.accessToken = null;
      this.refreshToken = null;
      this.user = null;
    }
    return result;
  }

  // ==================== PRODUCTS ====================

  async listProducts(params = {}) {
    return this._request('GET', '/products', null, { page: 1, limit: 20, ...params }, false);
  }

  async getProduct(id) {
    return this._request('GET', `/products/${id}`, null, null, false);
  }

  async createProduct(data) {
    return this._request('POST', '/products', data);
  }

  async updateProduct(id, data) {
    return this._request('PUT', `/products/${id}`, data);
  }

  async deleteProduct(id) {
    return this._request('DELETE', `/products/${id}`);
  }

  async getFeaturedProducts() {
    return this._request('GET', '/products/featured', null, null, false);
  }

  // ==================== SUPPLIERS ====================

  async listSuppliers(params = {}) {
    return this._request('GET', '/suppliers', null, { page: 1, limit: 20, ...params }, false);
  }

  async getSupplier(id) {
    return this._request('GET', `/suppliers/${id}`, null, null, false);
  }

  async getMySupplierProfile() {
    return this._request('GET', '/suppliers/me/profile');
  }

  async getSupplierStats(id) {
    return this._request('GET', `/suppliers/${id}/stats`, null, null, false);
  }

  // ==================== CATEGORIES ====================

  async getCategories() {
    return this._request('GET', '/categories', null, null, false);
  }

  async getCategoryBySlug(slug) {
    return this._request('GET', `/categories/slug/${slug}`, null, null, false);
  }

  // ==================== NEGOTIATIONS ====================

  async listNegotiations(params = {}) {
    return this._request('GET', '/negotiations', null, { page: 1, limit: 20, ...params });
  }

  async getNegotiation(id) {
    return this._request('GET', `/negotiations/${id}`);
  }

  async createNegotiation(productId, supplierId, message) {
    return this._request('POST', '/negotiations', {
      productId,
      supplierId,
      initialMessage: message
    });
  }

  async sendMessage(negotiationId, content, type = 'TEXT') {
    return this._request('POST', `/negotiations/${negotiationId}/messages`, { type, content });
  }

  async getMessages(negotiationId, page = 1) {
    return this._request('GET', `/negotiations/${negotiationId}/messages`, null, { page });
  }

  // ==================== HEALTH ====================

  async healthCheck() {
    try {
      const response = await fetch(this.baseUrl.replace('/api', '/health'));
      return { success: true, status: await response.text() };
    } catch (error) {
      return { success: false, error: error.message };
    }
  }

  async healthReady() {
    try {
      const response = await fetch(this.baseUrl.replace('/api', '/health/ready'));
      return await response.json();
    } catch (error) {
      return { success: false, error: error.message };
    }
  }
}

// ==================== TEST RUNNER ====================

class TestRunner {
  constructor() {
    this.client = new B2BApiClient();
    this.results = [];
    this.passed = 0;
    this.failed = 0;
  }

  printHeader(text) {
    console.log('\n' + '='.repeat(60));
    console.log(`  ${text}`);
    console.log('='.repeat(60));
  }

  printTest(name, success, details = '') {
    const status = success ? '✅ PASS' : '❌ FAIL';
    console.log(`${status} | ${name}`);
    if (details && !success) {
      console.log(`       └─ ${details}`);
    }
    
    this.results.push({ name, success, details });
    
    if (success) {
      this.passed++;
    } else {
      this.failed++;
    }
  }

  async runTest(name, testFunc) {
    try {
      const result = await testFunc();
      const success = result.success === true;
      const details = success ? '' : (result.error?.message || 'Unknown error');
      this.printTest(name, success, details);
      return result;
    } catch (error) {
      this.printTest(name, false, error.message);
      return { success: false, error: { message: error.message } };
    }
  }

  printSummary() {
    console.log('\n' + '='.repeat(60));
    console.log('  TEST SUMMARY');
    console.log('='.repeat(60));
    console.log(`  Total:  ${this.passed + this.failed}`);
    console.log(`  Passed: ${this.passed} ✅`);
    console.log(`  Failed: ${this.failed} ❌`);
    console.log('='.repeat(60));
  }

  // ==================== TEST SUITES ====================

  async testHealth() {
    this.printHeader('HEALTH CHECK');
    
    await this.runTest('Health Check', () => this.client.healthCheck());
    await this.runTest('Health Ready', () => this.client.healthReady());
  }

  async testAuth() {
    this.printHeader('AUTHENTICATION');
    
    // Login supplier
    let result = await this.runTest(
      'Login (Supplier)',
      () => this.client.login(CONFIG.supplier.email, CONFIG.supplier.password)
    );
    
    if (result.success) {
      await this.runTest('Get Current User', () => this.client.getMe());
      await this.runTest('Refresh Token', () => this.client.refreshTokens());
      await this.runTest('Logout', () => this.client.logout());
    }
    
    // Login shop
    await this.runTest(
      'Login (Shop)',
      () => this.client.login(CONFIG.shop.email, CONFIG.shop.password)
    );
    
    await this.client.logout();
    
    // Invalid login
    result = await this.runTest(
      'Login (Invalid - should fail)',
      () => this.client.login('invalid@email.com', 'wrongpassword')
    );
    
    if (!result.success) {
      this.passed++;
      this.failed--;
    }
  }

  async testProducts() {
    this.printHeader('PRODUCTS');
    
    let result = await this.runTest('List Products', () => this.client.listProducts({ limit: 5 }));
    
    const products = result.data || [];
    if (products.length > 0) {
      const productId = products[0].id;
      await this.runTest('Get Product Detail', () => this.client.getProduct(productId));
    }
    
    await this.runTest('Search Products', () => this.client.listProducts({ search: 'phone' }));
    await this.runTest('Get Featured Products', () => this.client.getFeaturedProducts());
    
    // Create product (requires auth)
    await this.client.login(CONFIG.supplier.email, CONFIG.supplier.password);
    
    const categories = await this.client.getCategories();
    const categoryId = categories.data?.[0]?.id;
    
    if (categoryId) {
      const testProduct = {
        name: `Test Product ${Date.now()}`,
        sku: `TEST-${Date.now()}`,
        shortDescription: 'Test product created by API test',
        description: 'Full description of test product',
        basePrice: 99.99,
        currency: 'USD',
        moq: 10,
        categoryId
      };
      
      const createResult = await this.runTest('Create Product', () => this.client.createProduct(testProduct));
      
      if (createResult.success) {
        const newProductId = createResult.data.id;
        
        await this.runTest('Update Product', () => this.client.updateProduct(newProductId, { basePrice: 149.99 }));
        await this.runTest('Delete Product', () => this.client.deleteProduct(newProductId));
      }
    }
    
    await this.client.logout();
  }

  async testSuppliers() {
    this.printHeader('SUPPLIERS');
    
    let result = await this.runTest('List Suppliers', () => this.client.listSuppliers());
    
    const suppliers = result.data || [];
    if (suppliers.length > 0) {
      const supplierId = suppliers[0].id;
      await this.runTest('Get Supplier Detail', () => this.client.getSupplier(supplierId));
      await this.runTest('Get Supplier Stats', () => this.client.getSupplierStats(supplierId));
    }
    
    await this.client.login(CONFIG.supplier.email, CONFIG.supplier.password);
    await this.runTest('Get My Profile', () => this.client.getMySupplierProfile());
    await this.client.logout();
  }

  async testCategories() {
    this.printHeader('CATEGORIES');
    
    const result = await this.runTest('Get Category Tree', () => this.client.getCategories());
    
    const categories = result.data || [];
    if (categories.length > 0 && categories[0].slug) {
      await this.runTest('Get Category by Slug', () => this.client.getCategoryBySlug(categories[0].slug));
    }
  }

  async testNegotiations() {
    this.printHeader('NEGOTIATIONS');
    
    await this.client.login(CONFIG.shop.email, CONFIG.shop.password);
    
    await this.runTest('List Negotiations', () => this.client.listNegotiations());
    
    const products = await this.client.listProducts({ limit: 1 });
    const suppliers = await this.client.listSuppliers();
    
    if (products.data?.length && suppliers.data?.length) {
      const product = products.data[0];
      const supplier = suppliers.data[0];
      
      const createResult = await this.runTest(
        'Create Negotiation',
        () => this.client.createNegotiation(
          product.id,
          supplier.id,
          "Hello, I'm interested in your product. What's the best price for 100 units?"
        )
      );
      
      if (createResult.success) {
        const negotiationId = createResult.data.id;
        
        await this.runTest('Get Negotiation Detail', () => this.client.getNegotiation(negotiationId));
        await this.runTest('Get Messages', () => this.client.getMessages(negotiationId));
        await this.runTest('Send Message', () => this.client.sendMessage(negotiationId, 'Looking forward to your response!'));
      }
    }
    
    await this.client.logout();
  }

  async runAll() {
    console.log('\n' + '🚀 '.repeat(20));
    console.log('  B2B MARKETPLACE API TEST SUITE');
    console.log('🚀 '.repeat(20));
    console.log(`\nBase URL: ${this.client.baseUrl}`);
    console.log(`Time: ${new Date().toISOString()}`);
    
    await this.testHealth();
    await this.testAuth();
    await this.testCategories();
    await this.testSuppliers();
    await this.testProducts();
    await this.testNegotiations();
    
    this.printSummary();
    
    return this.failed === 0;
  }
}

// ==================== MAIN ====================

async function main() {
  const args = process.argv.slice(2);
  
  if (args.includes('--url')) {
    const urlIndex = args.indexOf('--url');
    if (args[urlIndex + 1]) {
      CONFIG.baseUrl = args[urlIndex + 1];
    }
  }
  
  const runner = new TestRunner();
  
  if (args.includes('--auth')) {
    await runner.testHealth();
    await runner.testAuth();
  } else if (args.includes('--products')) {
    await runner.testHealth();
    await runner.testProducts();
  } else if (args.includes('--suppliers')) {
    await runner.testHealth();
    await runner.testSuppliers();
  } else if (args.includes('--negotiations')) {
    await runner.testHealth();
    await runner.testNegotiations();
  } else {
    const success = await runner.runAll();
    process.exit(success ? 0 : 1);
  }
  
  runner.printSummary();
}

main().catch(console.error);
